### ISSUE TITLE
Handle dormant and closed practices

### ISSUE THREAD

**Issue by sebbacon** [ebmdatalab/openprescribing#891]
A dormant practice is a practice that should not be doing any prescribing (typically because it's closing, but sometimes this is temporary and it becomes active again).  However, practices that are not actively prescribing **can** (and do) appear in prescribing data -- because prescribing data is recorded at the time a script is fulfilled by a dispenser, which can happen many months after the prescription is made (see #522 for more detail on this).

When a practice is dormant, we know it (at least sometimes) [can still move CCGs](https://github.com/ebmdatalab/openprescribing/issues/513#issuecomment-317433101).  As a specific example, `P84676` was in `00W` (Central Manchester) which merged with `14L` (Manchester) in April 2017.  The data for that month showed `14L` in the prescribing data but still showed `00W` in the practices table.

In our [recently-FOIed mergers data](https://github.com/ebmdatalab/openprescribing/issues/408#issuecomment-392811309), we can also see that `P84676` merged into `P84068` in October 2016; with data being assigned to it right until April 2017.

Per #401, #215, and #479, (all?) closed and (some?) dormant practices appear in data as assigned to old CCGs.

Suggested steps: 
* [ ] `epraccur` is quarterly, so gets out of date and should only be used for CCG membership as a fallback where no other data is available. Check that we use the **most recent** CCG membership for a practice, i.e. prefer membership information given in the latest month's prescribing data. The `P84676` example above suggests we do not do this consistently, or didn't a year ago.  XXX add a new issue and link to it from here. 
* Define an **orphan practice** as a closed or dormant practice which belongs to a CCG that no longer exists, even after (a) having applied available merger information, and (b) using the current month of prescribing data for CCG membership rather than the `epraccur` data (if the previous step suggests this is the right approach)
* Report on, and discuss, the impact of excluding orphan practices:
  * [ ] Make a report of "orphan" practices by month, going back 5 years, (a) any practices assigned to a closed CCG (#527 has handy resource listing these, should also be in our PCT model), (b) if we have merge information or not for each practice, (c) its status (closed, open, dormant). Assert that there are no open orphan practices.
  * [ ] Discuss the report, bearing in mind our merger data will, at least for the time being, get out of date (currently trying to establish the best way of getting the data monthly in the FOI linked above)
* If we think it's going to be OK to exclude orphan practices from our analyses, incorporate merger data into our dataset:
  * [ ] Assign all historic prescribing to the surviving practice, per our FOI spreadsheet
  * [ ] Make a note on the closing practice page describing where the patient list went (i.e. pointing to the surviving practice). This page should not show any analysis but a big link to the mergee practice.
  * [ ] Make a note on the surviving practice page listing mergee practices with names and dates of merged practices.
  * [ ] Make a note on a CCG page which no longer has any practices that it no longer exists. Consider adding a "notes" field which we display on that page to explain mergers or other custom information about the closure.
* Handle orphan data in analyses:
  * Where we don't have merger data, the patients do continue to exist, but we've lost them. 
  * [ ] At a CCG level this is OK as long as we still have a current CCG membership for them, as we can assume they've been incorporated elsewhere, even if we don't know where. However, if the assigned CCG no longer exists, we should exclude them, as attribution to a CCG which no longer exists is analytically useless
  * [ ] Decide what to do in practice-level measures and analyses with respect to orphan practices. I'm not sure if we should keep the practices in our outputs or not.  A dormant or closed practice can continue to dribble out prescriptions for months after they've closed, which will muck up ratios due to small numbers issues (see [this comment](https://github.com/ebmdatalab/openprescribing/issues/197#issuecomment-247301880) for a historic observation on a similar point). On the other hand, these are still real data which could contribute to overall centiles and analyses of historic interest  
  * [ ] If we decide to keep orphan practices, check we're not excluding them from measures. [This historic issue](https://github.com/ebmdatalab/openprescribing/issues/197) suggests at some point we were, but I don't think we are any longer.

Overall, we will know this is resolved correctly if the effects noted in #479, #401, #215, #513 and #408 are resolved.

**Comment by richiecroker** [comment]
As per #408:
@sebbacon i suspect the "effective date" is the one we want, as looking at it i think it refers to the date the merge actually happened - I suspect the "system merge date" is the date Capita merged the practices on the system (or not).

Either way, it shouldn't make too much difference, as we'll only see month granularity in the prescribing data.

**Comment by HelenCEBM** [comment]
Please can we update the [FAQ](https://openprescribing.net/faq/#gpclose) on this @evansd or @inglesp ?

It currently says:
> **A GP Practice has closed or merged - why is it included on your dashboards?**
OpenPrescribing uses the prescribing dataset published monthly by NHS Digital, which includes the past prescribing for practices that have merged or closed. Data will still show for a closed practice if a prescription prescribed by that surgery was dispensed in the community after the closure date.
The historical data for closed practices will continue to be shown on OpenPrescribing, and we are currently thinking about how we can indicate to the user that the practice no longer operates.

How about replacing the **last two lines** of the above with: 
> The historical data for closed practices will continue to be shown on OpenPrescribing, and any practice with a current status of Dormant or Closed will be labelled as such. They will continue to be shown under the CCG they were last recorded under

Here's the page to update:
 https://github.com/ebmdatalab/openprescribing/blob/a6a93e8500487071fc18ae0c07760f0e865a477f/openprescribing/templates/faq.html

**Comment by inglesp** [comment]
> Check that we use the **most recent** CCG membership for a practice, i.e. prefer membership information given in the latest month's prescribing data. The P84676 example above suggests we do not do this consistently, or didn't a year ago.

In the vast majority of cases where a practice has moved CCG in the last two years, we pick up the change in the April prescribing data and in the June `epraccur.csv` file.  Given that the April prescribing data is not available until June, and from what I can tell, the June `epraccur` data has always been published before we import the April prescribing data, I am confident that we do use the most recent CCG membership for a practice.

I have only found three practices where the`epraccur` data lags behind the prescribing data for a couple of months -- I assume this is an error from the publisher -- and I don't think it is worth taking any steps to mitigate this.

(These are `B86669`, `P84676`, and `Y04664`).

---

(Notes on how I did this follow.)

I loaded details of what we know about practice to CCG mappings from these two sources, going back to 2016_12 (we don't have older `epraccur.csv`s lying around) into a single database table with schema `CREATE TABLE IF NOT EXISTS mapping(source, month, practice, ccg)`.

I loaded the `epraccur` from the CSV files, and the prescribing data from the results of this BQ query:

```
SELECT
  "prescribing" AS source,
  pct AS ccg,
  practice,
  CONCAT(CAST(EXTRACT(YEAR FROM month) AS STRING), '_', FORMAT("%02d", EXTRACT(MONTH FROM month))) AS month
FROM
  hscic.prescribing
WHERE month >= TIMESTAMP("2016-12-01")
GROUP BY ccg, practice, month
ORDER BY practice, month
```

I then got rid of any practices that have not moved CCG with this SQL:

```
WITH
  grouped_practices AS (SELECT practice FROM mapping GROUP BY practice, ccg),
  moving_practices AS (SELECT practice FROM grouped_practices GROUP BY practice HAVING count(practice) > 1),
  setting4_practices AS (SELECT code FROM practices WHERE setting = '4')
INSERT INTO mapping1
SELECT * FROM mapping WHERE practice IN moving_practices AND practice IN setting4_practices;
```

(This also drops any practices that are not GP practices, ie those with `setting = '4'`.)

There are 669 practices which have been in more than one CCG.

```
c = sqlite3.connect('mapping.db')
sources = ['practice_details', 'prescribing']

sql = 'SELECT month, ccg FROM mapping1 WHERE practice = ? AND source = ? ORDER BY month'

data = {}

for practice in practices:
    results = {
        source: c.execute(sql, [practice, source]).fetchall()
        for source in sources
    }

    interesting = {
        source: [(r2[0], r1[1], r2[1]) for r1, r2 in zip(results[source], results[source][1:]) if r1[1] != r2[1]]
        for source in sources
    }

    key = tuple(
        (source, tuple(r[0] for r in interesting[source]))
        for source in sources
    )

    data[practice] = {
        'results': results,
        'interesting': interesting,
        'key': key,
    }

counter = Counter(practice_data['key'] for practice_data in data.values())

for k, c in counter.most_common():
    print(f'{c:4d} {k}')
```

This gives:

```
 513 (('practice_details', ('2018_06',)), ('prescribing', ('2018_04',)))
 125 (('practice_details', ('2017_06',)), ('prescribing', ('2017_04',)))
  19 (('practice_details', ('2018_06',)), ('prescribing', ()))
   7 (('practice_details', ('2017_06', '2018_06')), ('prescribing', ('2017_04', '2018_04')))
   1 (('practice_details', ('2017_08', '2018_06')), ('prescribing', ('2017_05', '2018_04')))
   1 (('practice_details', ('2017_06', '2018_06')), ('prescribing', ('2017_04',)))
   1 (('practice_details', ('2017_08',)), ('prescribing', ('2017_04',)))
   1 (('practice_details', ()), ('prescribing', ()))
   1 (('practice_details', ('2018_03', '2018_05', '2018_06')), ('prescribing', ('2017_12',)))
```

**Comment by inglesp** [comment]
I don't think the FOIed merger data is adequately in sync with our data for us to be able to do anything useful with it.

---

In the FOIed merger data, there are:

* 249 merged practices
   * but 175 are open according to our database
* 231 remaining practice
   * but 23 are closed according to our database

There are also a couple of practices that have apparently been merged multiple times:

* `Y01008` has apparently been merged into `P81085`, `P81011`, and `P81153`.  However, the three remaining practices are closed, while the merging practice is open.
* `A81003` has apparently been merged into `A81044` and `A81613`.  In this case, one of the remaining practices is open (A81044) but the other one and the merging practice is closed.

---

There are 55 GP practices in our database that belong to CCGs that are closed.

And of these, we only have merger data for 1 (one) of them.

The FOIed merger data also has a sheet listing 121 closed practices.  None of the 55 practices belong to closed CCGs are listed here.

---

The FOIed merger data goes back to 2015_12.  But 42 of 55 of the practices belonging to closed CCGs have a `closed_date` of 2015_12 or later.

**Comment by sebbacon** [comment]
Some other, specific examples arising from me also checking out the data:

* `G81006` is actively prescribing, but in the closures worksheet as having closed on 31/12/2016
* `E81626` is dormant in our main practices table, and we last saw prescribing in Nov 2017. It's in the closures worksheet as closing on 2017-06-30, before it became a mergee in 2017-08-01 (E81030, which according to the closures worksheet closed in 2016, but is still actively prescribing, merged into it on that date)

[SQL here](https://gist.github.com/sebbacon/107108d888d04aeb5c9fa2ceb67cdf9a#file-merging-sql-L1-L19)

**Comment by inglesp** [comment]
Here's a breakdown by CCG of the 55 GP practices in our database that belong to CCGs that are closed:

| Code  | Name | Close date | Count |
| - | - | - | - |
| 13P | NHS BIRMINGHAM CROSSCITY CCG | 2018-03-31 | 9 |
| 11H | NHS BRISTOL CCG | 2018-03-31 | 8 |
| 11T | NHS NORTH SOMERSET CCG | 2018-03-31 | 6 |
| 05P | NHS SOLIHULL CCG | 2018-03-31 | 5 |
| 02V | NHS LEEDS NORTH CCG | 2018-03-31 | 4 |
| 00W | NHS CENTRAL MANCHESTER CCG | 2017-03-31 | 4 |
| 03G | NHS LEEDS SOUTH AND EAST CCG | 2018-03-31 | 2 |
| 04X | NHS BIRMINGHAM SOUTH AND CENTRAL CCG | 2018-03-31 | 2 |
| 10H | NHS CHILTERN CCG | 2018-03-31 | 2 |
| 00F | NHS GATESHEAD CCG | 2015-03-31 | 2 |
| 10Y | NHS AYLESBURY VALE CCG | 2018-03-31 | 2 |
| 11C | NHS WINDSOR, ASCOT AND MAIDENHEAD CCG | 2018-03-31 | 2 |
| 10W | NHS SOUTH READING CCG | 2018-03-31 | 2 |
| 10N | NHS NORTH AND WEST READING CCG | 2018-03-31 | 1 |
| 01N | NHS SOUTH MANCHESTER CCG | 2017-03-31 | 1 |
| 12A | NHS SOUTH GLOUCESTERSHIRE CCG | 2018-03-31 | 1 |
| 03C | NHS LEEDS WEST CCG | 2018-03-31 | 1 |
| 10M | NHS NEWBURY AND DISTRICT CCG | 2018-03-31 | 1 |

From [earlier work](https://github.com/ebmdatalab/openprescribing/issues/891#issuecomment-403746198) I have a database table containing practices' CCG codes extracted from the `epraccur` data going back to 2016_12.

Looking first at `13P` (NHS BIRMINGHAM CROSSCITY CCG), we see that the 2018_06 `epraccur` data contains 16 records for practices in `13P`, with 179 records for practices that were in `13P` and now in `15E` (NHS BIRMINGHAM AND SOLIHULL CCG).

```
sqlite> WITH practices AS (
   ...>     SELECT DISTINCT(practice)
   ...>     FROM mapping
   ...>     WHERE ccg = '13P'
   ...> )
   ...> 
   ...> SELECT
   ...>     ccg,
   ...>     month,
   ...>     COUNT(month) AS num
   ...> FROM mapping
   ...> WHERE
   ...>     source = 'practice_details'
   ...>     AND practice IN practices
   ...> GROUP BY
   ...>     ccg,
   ...>     month
   ...> ORDER BY
   ...>     month,
   ...>     ccg
   ...> ;
```

ccg|month|num
-|-|-
13P|2016_12|194
13P|2017_03|195
13P|2017_06|195
13P|2017_08|195
13P|2017_12|195
13P|2018_03|195
13P|2018_05|195
13P|2018_06|16
15E|2018_06|179

These 16 records are all either closed (the 9 in the table above) or not GP practices.

I think we can assume that any prescribing data from all 16 practices can be considered to belong to `15E`, which is where all the other practices in the now-closed `13P` CCG have moved to.

I will check this for all the other closed CCGs.

**Comment by inglesp** [comment]
This mapping looks sensible.  There are a couple of CCGs that need manual checking (`04X`, `00F`).

old code | old name | new code | new name
-|-|-|-
00W | NHS CENTRAL MANCHESTER CCG | 14L | NHS MANCHESTER CCG
01N | NHS SOUTH MANCHESTER CCG | 14L | NHS MANCHESTER CCG
10H | NHS CHILTERN CCG | 14Y | NHS BUCKINGHAMSHIRE CCG
10Y | NHS AYLESBURY VALE CCG | 14Y | NHS BUCKINGHAMSHIRE CCG
10M | NHS NEWBURY AND DISTRICT CCG | 15A | NHS BERKSHIRE WEST CCG
10N | NHS NORTH AND WEST READING CCG | 15A | NHS BERKSHIRE WEST CCG
10W | NHS SOUTH READING CCG | 15A | NHS BERKSHIRE WEST CCG
11H | NHS BRISTOL CCG | 15C | NHS BRISTOL, NORTH SOMERSET AND SOUTH GLOUCESTERSHIRE CCG
11T | NHS NORTH SOMERSET CCG | 15C | NHS BRISTOL, NORTH SOMERSET AND SOUTH GLOUCESTERSHIRE CCG
12A | NHS SOUTH GLOUCESTERSHIRE CCG | 15C | NHS BRISTOL, NORTH SOMERSET AND SOUTH GLOUCESTERSHIRE CCG
11C | NHS WINDSOR, ASCOT AND MAIDENHEAD CCG | 15D | NHS EAST BERKSHIRE CCG
05P | NHS SOLIHULL CCG | 15E | NHS BIRMINGHAM AND SOLIHULL CCG
13P | NHS BIRMINGHAM CROSSCITY CCG | 15E | NHS BIRMINGHAM AND SOLIHULL CCG
02V | NHS LEEDS NORTH CCG | 15F | NHS LEEDS CCG
03C | NHS LEEDS WEST CCG | 15F | NHS LEEDS CCG
03G | NHS LEEDS SOUTH AND EAST CCG | 15F | NHS LEEDS CCG

---

Code:

```
import csv
import sqlite3

with open('ccgs.csv') as f:
    ccgs_by_code = {r['code']: r for r in csv.DictReader(f)}

c = sqlite3.connect('mapping.db')

sql = '''
WITH practices AS (
    SELECT DISTINCT(practice)
    FROM mapping
    WHERE ccg = ?
)

SELECT
    ccg,
    COUNT(ccg) AS num
FROM mapping
WHERE
    source = 'practice_details'
    AND practice IN practices
    AND month = ?
    AND ccg != ?
GROUP BY
    ccg
ORDER BY
    ccg
'''


table = []

# the month is the month of the first epraccur release after the CCG closed
for ccg, month in [
    ['13P', '2018_06'],
    ['11H', '2018_06'],
    ['11T', '2018_06'],
    ['05P', '2018_06'],
    ['02V', '2018_06'],
    ['00W', '2017_06'],
    ['03G', '2018_06'],
    # ['04X', '2018_06'],
    ['10H', '2018_06'],
    # ['00F', '2015_06'],
    ['10Y', '2018_06'],
    ['11C', '2018_06'],
    ['10W', '2018_06'],
    ['10N', '2018_06'],
    ['01N', '2017_06'],
    ['12A', '2018_06'],
    ['03C', '2018_06'],
    ['10M', '2018_06'],
]:
    rows = c.execute(sql, [ccg, month, ccg]).fetchall()
    rows = [r for r in rows if ccgs_by_code[r[0]]['org_type'] == 'CCG']

    assert len(rows) == 1
    row = rows[0]
    table.append([ccg, ccgs_by_code[ccg]['name'], row[0], ccgs_by_code[row[0]]['name']])


table.sort(key=lambda row: (row[2], row[0]))

for row in table:
    print(' | '.join(row))
```

**Comment by inglesp** [comment]
The situation for `04X` (NHS BIRMINGHAM SOUTH AND CENTRAL CCG) was confused by there being one practice (`M85124`) which moved from `04X` to `05L` (NHS SANDWELL AND WEST BIRMINGHAM CCG) in 2017_04.

All other practices in `04X` moved to `15E` (NHS BIRMINGHAM AND SOLIHULL CCG), and we should assume that the closed practices in `04X` did the same.

---

`00F` (NHS GATESHEAD CCG) closed in 2015_03, which is before we have easy access to `epraccur` data.  #408 states that it merged into `13T` (NHS NEWCASTLE GATESHEAD CCG).

**Comment by sebbacon** [comment]
Great, feels like we're getting there.  Bear in mind #527 describes an official file describing CCG changes; could be worth sanity checking against that.

In terms of implementation: should we maintain a table of "closed practice to CCG" mappings that we can be joined on in our `normalised_*` views?  That could include a "notes" column. Could possibly be maintained as a google sheet?  Could also perhaps make a note on the practice page that it was *last* in CCG X but we treat it as in CCG Y for CCG-level analyses?

In terms of developing a semi-automated (?) heuristic for deciding on CCG membership, it's looking like this, right? 

* When a CCG changes, count the number of practices that move from the old to the new one
* If all of them move from the old to the new one, assume closed practices in the old one should be assigned to the new one
* If one or two (?) moved CCGs, assume the same?

Questions:

* What about where boundaries and names change, but codes do not? (e.g. North Lancashire -> Morecombe Bay, as described in #527)
* What if several practices moved CCGs (which I'm guessing would appear to happen in the two boundary changes in #527)?

**Comment by inglesp** [comment]
> When do closed practices leave epraccur.csv?

Never!  We have epraccur data going back to, um, 2015_20, and each new file we have contains records for all practices that have been in previous files.

(However, we know that the CCGs of closed practices don't always get updated in the epraccur data, which is why we're doing all this work!)

**Comment by inglesp** [comment]
> When the boundary of 01H (was Cumbria, now North Cumbria) and 01K (was North Lancashire, now Morecambe Bay) changed, were there any orphan practices? What happened to them?

This question doesn't quite make sense: we've defined an orphan practice to be a closed practice whose CCG is closed.  But 01H and 01K never closed.

Instead, we could ask whether we think that there are any closed practices that should have changed CCG when the boundaries changed.

The boundary change happened in 2017_04, and 32 practices moved from 01H to 01K.  All of these practices are still active.

There are 11 practices still in 01H which are closed.  (And none which are dormant.)

Of these, 5 closed before 2017_04, and looking at the postcodes, I think the following should be moved into 01K:

* Y00715
* A82011
* A82619

For instance:

```
# select postcode from frontend_practice where code = 'A82011';
┌──────────┐
│ postcode │
├──────────┤
│ LA14 5SL │
└──────────┘

# select name, ccg_id, postcode from frontend_practice where postcode like 'LA14 %' and setting = 4 and status_code = 'A';
┌───────────────────────────┬────────┬──────────┐
│           name            │ ccg_id │ postcode │
├───────────────────────────┼────────┼──────────┤
│ DUKE STREET SURGERY       │ 01K    │ LA14 1LF │
│ ATKINSON HEALTH CENTRE    │ 01K    │ LA14 2LR │
│ BURNETT EDGAR MEDICAL CTR │ 01K    │ LA14 3HY │
│ THE FAMILY PRACTICE       │ 01K    │ LA14 5SL │
│ NORWOOD MEDICAL CENTRE    │ 01K    │ LA14 5ES │
│ ABBEY ROAD SURGERY        │ 01K    │ LA14 5JY │
└───────────────────────────┴────────┴──────────┘
```

**Comment by inglesp** [comment]
Now #975 has been deployed, I will:

- [ ] Run `handle_orphan_practices` for each consecutive pair of `epraccur.csv` files (turning off Slack notifications).  I expect this to pick up all, or almost all, of the changes 
- [ ] Make manual changes identified in https://github.com/ebmdatalab/openprescribing/issues/891#issuecomment-410751112
- [ ] Run `import_measures`
- [ ] Run `create_views`

**Comment by inglesp** [comment]
Before doing this, there were 55 orphan practices:

```
>>> Practice.objects.filter(setting=4, ccg__close_date__isnull=False).exclude(status_code='A').count()
55
```

We expect there to be 0 once the process is complete.

**Comment by inglesp** [comment]
Notes about the manual changes I've made are below:

### 00F

https://github.com/ebmdatalab/openprescribing/issues/891#issuecomment-403965318 says that the two practices in 00F should be changed to be in 13T.

### 13P

This CCG has now closed.  Between 2016_02 and 2016_06, 3 practices moved from 13P to 15E.  When the CCG closed, all active practices were also moved to 15E, and the management task automatically moved the closed practices.

### 07H

Between 2016_09 and 2016_12, 8 practices had their setting changed from 4 (GP practice) to 10 (hospital service).  This doesn't suggest there was a boundary change, and so no further action is required.

### 99A

Between 2016_09 and 2016_12, 6 practices had their setting changed from 4 to either 10 or 0 (other).  No boundary change.

### 01H

https://github.com/ebmdatalab/openprescribing/issues/891#issuecomment-410751112

### 05L

Between 2017_03 and 2017_06, 8 practices moved from 05L (NHS Sandwell and West Birmingham CCG) to 15E (NHS Birmingham and Solihull CCG).

Unfortunately, from looking at the postcodes, it doesn't appear that the practices are geographically clustered.  (Or if they are, we can't tell from the postcode.)

<details>
<summary>SQL</summary>

```
127.0.0.1 prescribing@prescribing=# SELECT DISTINCT(split_part(postcode, ' ', 1)) FROM frontend_practice WHERE code IN ('M85123', 'M85154', 'M91642', 'M85652', 'M85686', 'M85642', 'M85711', 'M85701');
┌────────────┐
│ split_part │
├────────────┤
│ B16        │
│ B17        │
│ B18        │
│ B19        │
│ B43        │
│ B6         │
└────────────┘
(6 rows)

Time: 0.959 ms
127.0.0.1 prescribing@prescribing=# SELECT count(*) FROM frontend_practice WHERE split_part(postcode, ' ', 1) IN (SELECT DISTINCT(split_part(postcode, ' ', 1)) FROM frontend_practice WHERE code IN ('M85123', 'M85154', 'M91642', 'M85652', 'M85686', 'M85642', 'M85711', 'M85701')) AND ccg_id = '05L';
┌───────┐
│ count │
├───────┤
│    34 │
└───────┘
(1 row)

Time: 1.867 ms
127.0.0.1 prescribing@prescribing=# SELECT count(*) FROM frontend_practice WHERE split_part(postcode, ' ', 1) IN (SELECT DISTINCT(split_part(postcode, ' ', 1)) FROM frontend_practice WHERE code IN ('M85123', 'M85154', 'M91642', 'M85652', 'M85686', 'M85642', 'M85711', 'M85701')) AND ccg_id = '15E';
┌───────┐
│ count │
├───────┤
│    25 │
└───────┘
(1 row)

Time: 3.026 ms
```
</details>

I've also plotted the practices in the two CCGs on a map:

![image](https://user-images.githubusercontent.com/28734/43910438-be827eca-9bf4-11e8-8a45-163b64481c37.png)

The green dots are the locations of moved practices; brown is 05L; blue is closed practices in 05L; grey is 15E; red is closed practices in 15E.

There are a handful of blue dot in predominantly grey territory, but I would not be confident in saying that they ought to have moved CCG.

As such, I don't think there's anything we can do here.

**Comment by inglesp** [comment]
> This is incorrect as 15E didn't exist then.

Of course, thanks.

> So in fact there is no problem here, because 15L is still open and it's OK for closed practices to remain assigned to it.

Yes, but this is similar to the boundary change between 01H and 01K, where we have decided to move closed practices from 01H to 01K, even though 01H is still open.

(When the boundary between 01H (was Cumbria, now North Cumbria) and 01K (was North Lancashire, now Morecambe Bay) changed, the practices that moved from 01H to 01K were in a clear geographic region, and so we've decided that any closed practices in that region should also be moved to 01K, even though 01H is still open.)

But I don't think there's anything we can do, so I'm going to close this.
